DAML-LF: Contract ID allocation scheme specification (#5784)
digital-asset/daml: daml-lf/spec/contract-id.rst

IDs before it hands out the transaction to the write service.

**Validation**: The allocation scheme commutes with transaction
projection. That is, the contract IDs for contracts created in the
projection of a transaction to a set of parties can be computed
solely from the projection and input seeds.

**Unlinkability**: It is computationally infeasible to link the contract
contents to the contract ID unless the create node is witnessed or the input seeds are known. The
contract contents include the contract instance, the template ID, the
stakeholders / signatories / maintainers, and the contract key.

**Freshness**: It is computationally infeasible to find contracts C1 and
C2 with the same contract ID such that one of the following holds:

* The contracts have different stakeholders.
* The contracts are created by different transactions.
* The contracts are created by different subactions of the same
  transaction forest where the seeds are pairwise different.

**Distinctness**: The ledger can enforce that the seeds are pairwise
different.

Contract Identifiers
^^^^^^^^^^^^^^^^^^^^

A *contract identifier* (or contract ID for short) is a sequence of
bytes defined as follows ::

>   ContractID := verisionPrefix ∥ discriminator ∥ suffix  

Reviewer: Should it be `versionPrefix`? If so, the same type happened further below again.

where

* ``∥`` is the concatenation operation; 
* ``verisionPrefix`` is 1 byte (equal to 0) used to version the
  contract ID scheme;
* ``discriminator`` is a sequence of 32 bytes. It is like a random
  UUID, but generated from an initial seed (called *submission seed*)
  via a fixed `derivation scheme <Allocation scheme for
  discriminators_>`_ that supports validation (see below).

> * ``suffix`` is a sequence of 0 to 95 bytes used to enforce global

Reviewer: Doesn't it say 96 in the DAML-LF 1 spec?

.. Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

V1 Contract ID allocation scheme
================================

Goals
^^^^^

* Allows ordering contract IDs and make DAML semantics depend on this
  order, e.g., for comparison builtin and maps using IDs as keys.
* Eliminate all contract ID translations for central commiter ledger

> 

Reviewer 2: ```suggestion
* Allows ledgers to store arbitrary information about the contract and the creating transaction in the contract ID if necessary.

```

Requirements
^^^^^^^^^^^^

**Executability**: DAML engine can determine ordering of the contract
IDs before it hands out the transaction to the write service.

**Validation**: The allocation scheme commutes with transaction
projection. That is, the contract IDs for contracts created in the
projection of a transaction to a set of parties can be computed

> solely from the projection and input seeds.

Reviewer 2: This is the first time in this document that you mention "input seeds". Can you explain them briefly here or put a forward reference to their definition?

Author: I drop the "input seeds". I just consider the seed of the node to be part of the projection.

**Unlinkability**: It is computationally infeasible to link the contract

> contents to the contract ID unless the create node is witnessed. The

Reviewer 2: As is this is too strong. Given that transactions are small in size, an adversary that knows just the input seed used for the transaction that creates a contract could derive seeds for all potential nodes in the tree and then validate its guess of the signatories by recomputing the discriminators.
```suggestion
contents to the contract ID unless the create node is witnessed or the input seeds are known. The
```

contract contents include the contract instance, the template ID, the
stakeholders / signatories / maintainers, and the contract key.

**Freshness**: It is computationally infeasible to find contracts C1 and
C2 with the same contract ID such that one of the following holds:

* The contracts have different stakeholders.
* The contracts are created by different transactions.
* The contracts are created by different subactions of the same
  transaction forest where the seeds are pairwise different.

**Distinctness**: The ledger can enforce that the seeds are pairwise
different.

Contract Identifiers
^^^^^^^^^^^^^^^^^^^^

A *contract identifier* (or contract ID for short) is a sequence of
bytes defined as follows ::

  ContractID := verisionPrefix ∥ discriminator ∥ suffix  

where

* ``∥`` is the concatenation operation; 

> * ``verisionPrefix`` is 1 byte (equal to 0) used to versioned the

Reviewer 2: ```suggestion
* ``verisionPrefix`` is 1 byte (equal to 0) used to version the
```

  contract ID scheme;

> * ``discriminator`` is a sequence of 32 bytes. It is like a random
>   UUID, but generated from an initial seed (called *submission seed*)

Reviewer 2: ```suggestion
* ``discriminator`` is a sequence of 32 bytes. It is generated from an initial seed (called *submission seed*)
```
I'd try to avoid the technical term "UUID" because there's a standard that describes the precise format. In particular, a UUID consists of exactly 16 bytes.

  via a fixed `derivation scheme <Allocation scheme for
  discriminators_>`_ that supports validation (see below).
* ``suffix`` is a sequence of 0 to 95 bytes used to enforce global

>   uniqueness of the contract ID in a distributed ledger.

Reviewer 2: Is this enforced in the implementation anywhere? In our tests, even very long suffixes seem to be accepted: https://github.com/DACH-NY/canton/blob/master/common/src/test/scala/com/digitalasset/canton/protocol/ContractIdTest.scala#L70

Author: It is specified in the [DAML-LF spec](https://github.com/digital-asset/daml/blob/remy-contract-id-spec/daml-lf/spec/daml-lf-1.rst) line 564 (with a typo :().
I will fix the typo and add a check in a upcoming PR.

Discriminator freshness
-----------------------

> In a transaction we distinct two kinds of contract IDs:

Reviewer 2: ```suggestion
In a transaction we distinguish two kinds of contract IDs:
```

* The *local* contract IDs are the IDs of the contracts created by the
  transaction.

> *  The *global* contract IDs are the contract IDs that:

Reviewer 2: ```suggestion
* The *global* contract IDs are the contract IDs that:
```

>    * appears in the commands that produced the transaction. This

Reviewer 2: ```suggestion
   * appear in the commands that produced the transaction. This
```

     includes the IDs of the exercised contract, together with all the
     IDs referenced in the arguments of the create and exercise
     commands;

>    * that are fetched or looked up by key;

Reviewer 2: ```suggestion
   * that are fetched or looked up by key unless they are local;
```
A local contract may be the result of a `fetchByKey`, but we don't want it to be global, do we?

   * are referenced in the input contracts.

Note that local contract IDs correspond to the ID of output contracts
together with those contracts that have been created and archived in
the transaction. On the other hand, global contract IDs do not
reference only IDs of some contracts that have been previously
persisted on the ledger, but also any arbitrary contract IDs that the

> submitter have referenced in its submission.

Reviewer 2: ```suggestion
submitter has referenced in its submission.
```

The so-called *discriminator freshness condition* holds if the
discriminators from local contract IDs are distinct from the
discriminators from global contract IDs.

This ensures that only the discriminators, not the suffix are needed

> to order the contract IDs of created contract w.r.t. other contract

Reviewer 2: ```suggestion
to order the contract IDs of created contracts w.r.t. other contract
```

IDs. 

Contract ID uniqueness
----------------------

During interpretation local contract IDs are created without suffix.

> It is the responsibility of the ledger implementation to provide for
> each local contract ID a suffix that guarantee global uniqueness of
> the whole contract ID. No other requirement (except the 95 bytes size
> limit) is assumed for those suffices.

Reviewer 2: I think this does not really cover Corda's use of suffixes, namely storing additional information in the contract ID. How about the following?

```suggestion
Ledger implementations are responsible for enforcing 
uniqueness of contract IDs on the whole ledger.
This can be done by enforcing global uniqueness of the seeds or by appropriately suffixing the contract IDs.
No other requirement (except the 95 bytes size
limit) is assumed for those suffices.
```

> This guard against the submitter choosing low-entropy submission seeds. 

Reviewer 2: Suffixes do not guard against lack of entropy, because the low entropy remains in the discriminators. So unlinkability remains broken if entropy is low, even with suffixes. I suggest to remove this sentence.

Author: dropped

The simplest approach consists to suffix all local contract ID with a
uniquely global transaction ID. Alternatively central commiter ledger
can completely avoid suffixing by enforcing that same pair (submission
seed, submission time) is never used more that once, the discriminator
allocation scheme ensuring in this case the uniqueness of allocated
discriminators.

Submission time
---------------

The submission time is used to derive the transaction seed. In

> practice, it should be closed to the original time at which the

Reviewer 2: ```suggestion
practice, it should be close to the original time at which the
```

submission was initiated. No particular requirement is made in this
specification. However a particular ledger implementation can require
this time to be within a particular time window around the commit
time. This allows, for instance, the commiter of a central commiter

> ledger to en enforce global uniqueness of the pair (submission seed,

Reviewer 2: ```suggestion
ledger to enforce global uniqueness of the pair (submission seed,
```

submission time).

Allocation scheme for discriminators
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
From an initial seed, called in the following *submission seed*, seeds

> for all nodes in the transaction tree are derived as they are

Reviewer 2: ```suggestion
for all Create and Exercise nodes in the transaction tree are derived as they are
```
Fetch and LookupByKey don't really get seeds, do they?

computed.

We propose the following scheme, inspired by NIST’s HMAC_DRBG
construction. It assumes a hash-based MAC function HMAC, and defines

> for each big-endian 32 bits integer ``n`` the derived seed::

Reviewer 2: ```suggestion
for each big-endian 32 bits integer ``n`` the derived seed:
```

  deriveSeed(seed, n) := HMAC(seed, n)
  

Submission seed
---------------

Any random number with sufficient entropy works as submission seed
``submission_seed`` (256 bit entropy).

Transaction seed construction
-----------------------------

From the submission, a so-called *transaction seed* is derived as follows::

>   deriveTransactionSeed(submissionSeed, participandId, submissionTime, submitters) :=

Reviewer 2: `Engine.deriveTransactionSeed` doesn't include the submitters. https://github.com/digital-asset/daml/blob/master/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala#L391 So maybe remove them here too?

Author: good catch. 
It is removed.

>      HMAC(submissionSeed, participantId ∥ submissionTime ∥ nₛ ∥ submitters)

Reviewer 2: `participantId` is a ledger string, i.e., of varying length. In the above notation, I do not see that the length of `participantId` is included in the argument to `HMAC`. If it isn't, we may get hash collissions because `submitters` too is a variable-length string. In the implementation I see that the actual derivation adds the length of the participantId to the string. Can you also do so here?

Reviewer 2: Found the comment below about the length prefix.

where

* ``submissionSeed`` is the submission seed;
* ``participandId`` is the participant ID;
* ``submissionTime`` is the submission time in micro second encoded as
  a 64 bytes big-endian integer;
* ``nₛ`` is the number of submitters encoded as a 32 big-endian bits
  integer;
* ``submitters`` is the concatenation of the submitter IDs sorted
  lexicographically.

Strings such as participant IDs or submitter IDs are interpreted as

> their US-ASCII encoding prefixed with there size encoded as a

Author: ```suggestion
their US-ASCII encoding prefixed with their size encoded as a
```

> 32 bits big-endian integer.

Reviewer 2: Is the size prefix included when the submitters are sorted lexicographically?

Author: Yes. always you get collision for sumittes (`a`, `aaaa`) and (`aa, `aaa`).

Concretely  (`a`, `aaaa`) is encoded as `(2 ∥ 1 ∥ a ∥ 4 ∥ aaaa)` and  (`aa, `aaa`) as  `(2 ∥ 2 ∥ aa ∥ 3 ∥ aaa)`

For each root node of the transaction, a root seed `rootSeedᵢ` is
computed from the initial seed as follows::

  rootSeedᵢ = deriveSeed(transactionSeed, i)

Derivation of seeds for the children of exercise nodes
------------------------------------------------------

For an exercise node with seed ``nodeSeed``, the seeds `childSeedᵢ` for the
children are derived as follows::

  childSeedᵢ = deriveSeed(nodeSeed, i)

Derivation of contract ID discriminator
---------------------------------------

The discriminator for the created contract ID is given as follows::

  discriminator = HMAC(nodeSeed, submissionTime ∥ nₛ ∥ stackholders) 

where

* ``nodeSeed`` is the seed of the node where the contract is created;
* ``submissionTime`` is the submission time in micro second encoded as
  a 64 bytes big-endian integer;
* ``nₛ`` is the number of stakeholder's of the contract encoded as a
  32 bits big-endian integer;
* ``stackholders`` is the concatenation of the stakeholders IDs sorted
  lexicographically.

IDs of stakeholder are interpreted as their US-ASCII encoding prefixed
with there size encoded as a 32 bits  big-endian integer.
  
Submission
^^^^^^^^^^

> The submission the following steps:

Reviewer 2: ```suggestion
The submission performs the following steps:
```
Or some other verb that you'd like to have in there.

> * Peek a submission seed  with high entropy.

Reviewer 2: ```suggestion
* Pick a submission seed with high entropy.
```

* Derive the transaction seed and start the interperation
* During interpretation derive the node seeds, and the discriminator
  of local contract ID accordingly the scheme described above.  If the
  discriminator is not `fresh <Discriminator Freshness_>`_, abort the

>   interpretation. The submitter can restart the interpretation with
>   another submission seed.

Reviewer 2: According to step 1, the submission seed is picked by "submission". This sounds as if the submitter does not have control over the seed and therefore. This reads as if the submitter could choose the submission seed. So I recommend to either change the first step into `Submission takes a high-entropy submission seed as input` or to change this into 
```suggestion
  interpretation. The submitter can restart the interpretation, which will pick
  another submission seed.
```

* If the transaction succeeds, the output is a *raw transaction*

> * All local contract IDs in the raw transaction are suffix with a

Reviewer 2: ```suggestion
* All local contract IDs in the raw transaction are suffixed with a
```

  ledger specific suffix. This yields the *ready transaction*.
* The ready transaction is then sent on the write path, along with the

>   submission seed and the submission time.

Reviewer 2: What is the write path? I presume that this is something else than the Write service, as the suffixing only happens after the transaction has been sent to the Write service.

Author: I reformulated.

Validation
^^^^^^^^^^

Reinterpretation for a full transaction validation takes the

> transition, the submission seed, and the submission time as

Reviewer 2: ```suggestion
transaction, the submission seed, and the submission time as
```
